File: demos/demo_texture.cpp
/*******************************************************************************
 * Demo: Textured Ground and Triangle
 *
 * Demonstrates loading PNG textures and applying them to geometry.
 * The demo falls back gracefully to solid colours if texture files are missing.
 *
 * Texture files expected (relative to the build directory at run time):
 *   ../data/textures/triangle.png  — applied to the triangle
 *
 * Pattern:
 *   GLuint tex = app.getRenderer().loadTexture("path/to/file.png");
 *   scene.Add(triangle, ..., textureHandle=tex);
 *
 * loadTexture() returns 0 if the file cannot be opened; passing 0 as the
 * textureHandle causes the geometry to render with its solid colour instead,
 * so no special fallback logic is required in the scene update.
 *
 * Note: this demo uses a manual draw callback (app.addDrawCallback) rather than
 * app.addScene() because the textured Add() overloads are only available when
 * rendering is driven explicitly through the Renderer.
 *******************************************************************************/

#include "geodraw/app/app.hpp"
#include "geodraw/scene/scene.hpp"
#include "geodraw/geometry/geometry.hpp"
#include <iostream>

using namespace geodraw;

struct DemoGeometry {
  Pose3 axes;
  // Note: UV coordinates are set to enable texture sampling.
  Triangle3 triangle{Pos3{1, 1, 0}, Pos3{3, 2, 0}, Pos3{1, 2, 1},
                     glm::vec2{0.0f, 0.0f}, glm::vec2{1.0f, 0.0f}, glm::vec2{0.5f, 1.0f}};
  // Load textures through the renderer.hpp
  // Returns 0 if the file cannot be opened — the demo still runs without them.
  GLuint triangleTex;
};

void demoGeometryInit(App& app, DemoGeometry& geometry) {
  geometry.triangleTex = app.getRenderer().loadTexture("../data/textures/brick.png");
  if (geometry.triangleTex == 0) {
    std::cout << "Note: brick.png not found — triangle will be solid orange.\n";
  }
}

void UpdateDemoGeometry(Scene &scene, DemoGeometry &geometry) {
  scene.clear();
  scene.Add(geometry.axes, /*scale=*/2.0f, /*thickness=*/5.0f);
  scene.Add(geometry.triangle, /*thickness=*/2.0f, /*linestyle=*/LineStyle::Solid,
  /*alpha=*/1.0, /*color=*/Colors::ORANGE, /*depthTest=*/true, /*filled=*/true, /*textureHandle=*/geometry.triangleTex);
}

int main() {
    App app(1280, 720, "Demo Textured Geometry");
    app.addScene();
    DemoGeometry geometry;
    demoGeometryInit(app, geometry);

    app.addUpdateCallback([&](float /*dt*/) {
      UpdateDemoGeometry(app.scene(), geometry);
    });

    app.addDrawCallback([&]() {
      app.getRenderer().render(app.scene(), app.camera);
    });

    app.run();

    // Release textures when done.
    if (geometry.triangleTex != 0) app.getRenderer().unloadTexture("../data/textures/brick.png");

    return 0;
}

File: python/demos/demo_texture.py
"""Demo: Textured Geometry (Python)

Python mirror of app/demo_texture.cpp.

Demonstrates loading PNG textures and applying them to geometry.
The demo falls back gracefully to solid colours if texture files are missing.

Texture files expected (relative to the repo root at run time):
  data/textures/brick.png — applied to the triangle

Pattern:
  tex = app.renderer().load_texture("path/to/file.png")  # returns int handle, 0 on error
  scene.add_triangle(vertices, ..., texture_handle=tex)

load_texture() returns 0 if the file cannot be opened; passing 0 as the
texture_handle causes the geometry to render with its solid colour instead,
so no special fallback logic is required in the scene update.

Triangle UV coordinates are required for texture sampling:
  vertices_uv = [
      ((1, 1, 0), (0.0, 0.0)),
      ((3, 2, 0), (1.0, 0.0)),
      ((1, 2, 1), (0.5, 1.0)),
  ]

Run from anywhere:
    python python/demos/demo_texture.py
"""

import os
import geodraw

_REPO_ROOT = os.path.normpath(os.path.join(os.path.dirname(os.path.realpath(__file__)), "..", ".."))
_BRICK_TEX = os.path.join(_REPO_ROOT, "data", "textures", "brick.png")


def main():
    app = geodraw.App(1280, 720, "Demo Textured Geometry")

    tex = app.renderer().load_texture(_BRICK_TEX)
    if tex == 0:
        print("Note: brick.png not found — triangle will be solid orange.")

    # UV coordinates enable texture sampling.
    TRIANGLE_UV = [
        ((1, 1, 0), (0.0, 0.0)),
        ((3, 2, 0), (1.0, 0.0)),
        ((1, 2, 1), (0.5, 1.0)),
    ]
    ORANGE = (0.95, 0.5, 0.0)

    @app.add_update_callback()
    def update(dt):
        scene = app.scene()
        scene.clear()
        scene.add_axes((0, 0, 0), scale=2.0, thickness=5.0)
        scene.add_triangle(TRIANGLE_UV, color=ORANGE, texture_handle=tex)

    app.run()

    if tex != 0:
        app.renderer().unload_texture(_BRICK_TEX)


if __name__ == "__main__":
    main()
